### ISSUE TITLE
BBK* throws null pointer exception

### ISSUE THREAD

**Issue by gusennan** [donaldlab/OSPREY3#108]
BBK* throws a NullPointerException on prepped 5aei.  It's unclear if this is caused by a bug in the implementation of the algorithm, a programming error, or something else.  The stack trace of the failed run is below, and the input structure and python script are attached.

[bbkstar-bug.zip](https://github.com/donaldlab/OSPREY3/files/4261710/bbkstar-bug.zip)

```
➜  florian python 5aei_kstar.py 
OSPREY 3.1-beta1-b162
Using up to 30000 MiB heap memory: 10048 MiB for garbage, 19952 MiB for storage
read PDB file from file: 5aei_final_osprey-noWater_mut-Tyr-binder.pdb
### Started: edu.duke.cs.osprey.kstar.BBKStar$ConfSpaceInfo@7853363c
Calculating reference energies for 125 residue confs...
Progress:  100.0%   ETA: 0 ns
Finished in 172.0 ms
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/emat.protein.dat
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/emat.protein.rigid.dat
### Started: edu.duke.cs.osprey
.kstar.BBKStar$ConfSpaceInfo@fb5d334
Calculating reference energies for 91 residue confs...
Progress:  100.0%   ETA: 0 ns
Finished in 43.2 ms
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/emat.ligand.dat
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/emat.ligand.rigid.dat
### Started: edu.duke.cs.osprey.kstar.BBKStar$ConfSpaceInfo@45dc7be
Calculating reference energies for 216 residue confs...
Progress:  100.0%   ETA: 0 ns
Finished in 91.6 ms
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/emat.complex.dat
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/emat.complex.rigid.dat
computing K* score for the wild-type sequence...
Making energy matrix for edu.duke.cs.osprey.energy.ConfEnergyCalculator@3f64d943
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/protein.default.emat
Making energy matrix for edu.duke.cs.osprey.energy.ConfEnergyCalculator@4a0c3aa8
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/ligand.default.emat
Making energy matrix for edu.duke.cs.osprey.energy.ConfEnergyCalculator@26a33cf
read energy matrix from file: /usr/project/dlab/Users/nguerin/collaborations/florian/complex.default.emat
Total Z upper bound reduction through minimizations: -4.114647e+09
Average Z upper bound reduction per minimizations: -5.143308e+08
Total Z upper bound reduction through minimizations: 1.098857e-01
Average Z upper bound reduction per minimizations: 7.229325e-04
Total Z upper bound reduction through minimizations: 1.539918e+49
Average Z upper bound reduction per minimizations: 1.924897e+48
sequence    0/   0   159=glu 162=val   K*(log10): 25.558831 in [25.157639,25.916724] (log10)   protein: [22.706870,22.715215] (log10)                   , numConfs: 8, epsilon: 0.019   ligand: [-0.208957, 0.183889] (log10)                   , numConfs: 152, epsilon: 0.595   complex: [48.056743,48.414637] (log10)                   , numConfs: 8, epsilon: 0.561
computing K* scores for the 3 best sequences to epsilon = 0.6 ...
Refining sequence 
Traceback (most recent call last):
  File "5aei_kstar.py", line 93, in <module>
    scoredSequences = bbkstar.run()
jpype._jexception.NullPointerExceptionPyRaisable: java.lang.NullPointerException

java.lang.NullPointerException
        at edu.duke.cs.osprey.kstar.BBKStar$MultiSequenceNode.calcUpperBound(BBKStar.java:335)
        at edu.duke.cs.osprey.kstar.BBKStar$MultiSequenceNode.estimateScore(BBKStar.java:255)
        at edu.duke.cs.osprey.kstar.BBKStar.run(BBKStar.java:641)
```

**Comment by cuchaz** [comment]
The offending line of code is this one:
```java
info.confSearchFactoryMinimized.make(rcs),
```
which I think means this might be a MARK*/BBK* interaction bug? If I remember correctly, MARK* doesn't use the conf search factories, so `info.confSearchFactoryMinimized` would always be null?

**Comment by eternityslyre** [comment]
This is quite curious, since I run BBK*+MARK* all the time. I'm running it
off of master right now for the SARS-CoV-2 designs. I'll try to take a look
this weekend.

**Comment by cuchaz** [comment]
Hmm... It look like the bug happened on build 162. The latest build is 165. If the master branch does indeed work, what are the chances this bug was already fixed in one of the recent builds? Or maybe the bug isn't fixed yet, but doesn't happen on all designs somehow.

**Comment by cuchaz** [comment]
Ok, the last three builds are mine, and I didn't touch BBK*/MARK* code, so probably this bug still happens in the master branch.

**Comment by eternityslyre** [comment]
That's the likeliest scenario. If I don't find something in the next hour
or so I'll dig into it later.

**Comment by cuchaz** [comment]
I'm still trying to triage this bug, but I'm not an expert on how MARK* uses the BBK* code... It looks like the `SingleSequenceNode` class has been updated to use the new `pfuncFactory` field, but the `MultiSequenceNode` class has not. From poking the code a bit, If the code ever scores a multi-sequence node, I think that will cause the crash. But for designs with a small number of mutable residues (ie 2), we never compute scores for multi-sequence nodes, and the bug doesn't happen?

**Comment by eternityslyre** [comment]
That appears to be the issue. I've pushed a fix that automatically creates
confsearchfactories. It'll require setting confECalcRigid, a new variable,
in the ConfSpaceInfo class.

**Comment by jonathanjou** [comment]
Thanks for managing this bug, everyone. I apologize for the rushed
solution. We'll eventually need to properly divorce BBK* from reliance on
ConfSearchFactories, given that SHARK* contains a better multisequence
bound. Hopefully modifying the following line (in bold, below) in our
example scripts will resolve the issue.

# configure BBK* inputs for each conf space
for info in bbkstar.confSpaceInfos():
	print ("### Started: {}".format(info))
	# how should we define energies of conformations?
	eref = osprey.ReferenceEnergies(info.confSpace, minimizingEcalc)
	info.confEcalcMinimized = osprey.ConfEnergyCalculator(info.confSpace,
minimizingEcalc, referenceEnergies=eref)
	# compute the energy matrix
	emat = osprey.EnergyMatrix(info.confEcalcMinimized,
cacheFile='emat.%s.dat' % info.id)
	# BBK* needs rigid energies too*	info.confEcalcRigid =
osprey.ConfEnergyCalculatorCopy(info.confEcalcMinimized, rigidEcalc)*
	rigidEmat = osprey.EnergyMatrix(rigidConfEcalc,
cacheFile='emat.%s.rigid.dat' % info.id)
	# how should partition functions be computed?
	info.pfuncFactory = osprey.PartitionFunctionFactory(info.confSpace,
info.confEcalcMinimized, info.id)

**Comment by gusennan** [comment]
Heya,

I saw the [commit made on Friday](https://github.com/donaldlab/OSPREY3/commit/6c04e51340a99333b7151f64bb60e57e8bb44103) and adapted Florian's script to include this line.  It doesn't appear to fix the bug. Specifically:

1. `info.confEcalcRigid =
osprey.ConfEnergyCalculatorCopy(info.confEcalcMinimized, rigidEcalc)` means that there's no `rigidConfEcalc` defined anymore, which is used on the following line. 
2. If I substitute `info.confEcalcRigid` for `rigidConfEcalc` on the following line, I still get a NPE.

```
Traceback (most recent call last):
  File "5aei_kstar.py", line 93, in <module>
    scoredSequences = bbkstar.run()
jpype._jexception.NullPointerExceptionPyRaisable: java.lang.NullPointerException

java.lang.NullPointerException
        at edu.duke.cs.osprey.kstar.BBKStar$MultiSequenceNode.calcUpperBound(BBKStar.java:335)
        at edu.duke.cs.osprey.kstar.BBKStar$MultiSequenceNode.estimateScore(BBKStar.java:255)
        at edu.duke.cs.osprey.kstar.BBKStar.run(BBKStar.java:641)
```

**Comment by jonathanjou** [comment]
Hm. I'll take a look tomorrow. Should be easy enough to fix.

**Comment by eternityslyre** [comment]
Thanks for checking, Nate. It looks like line 335 of the current
BBKStar.java doesn't correspond to calcUpperBound anymore, which suggests
that the error is being generated by a version OSPREY from before my
commit. If you're still running into issues after updating we can look over
it tomorrow.

**Comment by gusennan** [comment]
Closing, this was fixed by d931e0b1675a4c921b88bacb309a0b046fd9d3e5 and 6c04e51340a99333b7151f64bb60e57e8bb44103
